articles/20230820_dev-container-vscode.md

```diff
@@ -0,0 +1,113 @@
+- **Flexibility**: You can choose a base image with the software and tools you want or build your own custom base image, meeting your specific needs.
+
+Dev containers revolutionize the developer experience by providing pre-configured, isolated environments that can supercharge your productivity. If you haven't tried them yet, you owe it to yourself as a developer to give dev containers a shot. They may just change the way you work! The future is containerized!
+
+## Prerequisites
+
+Before you begin, ensure you have the following:
+
+- **Docker**: Installed and running on your machine.
+- **VSCode**: Installed with the Remote - Containers extension.
+- **Basic Familiarity**: With Docker and VSCode.
+
+## Step 1: Create a `Dockerfile`
+
+A Dockerfile specifies the instructions to create the image for your development environment. Follow these steps:
+
+1. **Create a Directory**: For your project and navigate into it.
+
+   ```sh
+   mkdir my-dev-container && cd my-dev-container
+
+   ```
+
+2. **Create a Dockerfile**: Inside the directory, create a file named Dockerfile with the following content:
+
+   ```sh
+   FROM ubuntu:20.04
+
+   # Avoid warnings by switching to noninteractive
+   ENV DEBIAN_FRONTEND=noninteractive
+
+   # Use the default user 'developer'
+   ARG USERNAME=developer
+   ARG USER_UID=1001
+   ARG USER_GID=$USER_UID
+
+   # Create the user 'developer' with sudo access
+   RUN groupadd --gid $USER_GID $USERNAME \
+       && useradd --uid $USER_UID --gid $USER_GID -m $USERNAME \
+       && apt-get update \
+       && apt-get install -y sudo \
+       && echo $USERNAME ALL=\(root\) NOPASSWD:ALL > /etc/sudoers.d/$USERNAME \
+       && chmod 0440 /etc/sudoers.d/$USERNAME
+
+   # Install development tools
+   RUN apt-get install -y git build-essential
+
+   # Switch back to the dialog frontend for any additional apt-get installations
+   ENV DEBIAN_FRONTEND=dialog
+   ```
+
+## Step 1: Create a `devcontainer.json` File
+
+The devcontainer.json file describes how VSCode should interact with the development container. In the same directory, create a file named devcontainer.json with the following content:
+
+    ```sh
+    {
+    "name": "My Development Container",
+    "build": {
+        "dockerfile": "Dockerfile",
+        "args": {
+        "USER_UID": "1001",
+        "USER_GID": "1001"
+        }
+    },
+    "remoteUser": "developer"
+    }
+
+This file tells VSCode to build the development container image using the Dockerfile in the current directory and set the remote user to `developer`.
+
+### Step 3: Open the Project in VSCode
+
+1. **Open the Directory**: You created in VSCode.
+2. **Reopen in Container**: Press `F1` to open the command palette and select `Remote-Containers: Reopen in Container`. This will build the Docker image if it's not already built and start a container with your development environment, attaching VSCode to it.
+
```

Comment:
Add Step 3b. Open the Project in Daytona

Explain why and how in same amount of text as above.